src/x11_utils.rs

```diff
@@ -37,6 +37,47 @@ pub trait TryParse: Sized {
     fn try_parse(value: &[u8]) -> Result<(Self, &[u8]), ParseError>;
 }
 
+/// A representation of the header of a request.
+#[derive(Debug, Clone, Copy)]
+pub struct RequestHeader {
+    /// The major opcode of the request.
+    pub major_opcode: u8,
+    /// The minor opcode of the request (which, for some requests, may not be an
+    /// opcode at all).
+    pub minor_opcode: u8,
+    /// The remaining length of the request, in units of 4 bytes. Unlike the wire format,
+    /// this does *not* include the header itself. If the BigRequests extension is enabled
+    /// this can be greater than u16::max_value - 1.
+    pub remaining_length: u32,
+}
+
+/// Validates that the header and value lengths match and that the major (for xproto requests)
+/// or minor (for extension requests) match what is expected. For extension requests it is the
+/// responsibility of the try_parse_request caller to validate the major opcode.
+pub(crate) fn validate_request_pieces(
+    header: RequestHeader,
+    value: &[u8],
+    expected_major_opcode: Option<u8>,
+    expected_minor_opcode: Option<u8>,
+) -> Result<(), ParseError> {
+    if header.remaining_length as usize * 4 != value.len() {
```

Comment:
This is not necessary. If you want to see something semi-scary, look at `src/cairo-xlib-surface.c` in cairo's source code. The function `_x_bread_crumb` is printf-like and sends a string as an over-long `NoOperation` request.

I think this length check can just be completely removed. I haven't seen the later commits yet, but I would guess that some code elsewhere uses the "on-the-wire request length" to construct a slice of the right length on `value`. If something tries to read past the end of that slice, `TryParse` should turn that into a `ParseError`.